origin-js/src/resources/marketplace.js

```diff
@@ -69,11 +138,20 @@ export default class Marketplace {
       return listingIds
     }
 
-    return Promise.all(
-      listingIds.map(async listingId => {
-        return await this.getListing(listingId)
-      })
-    )
+    if (opts.withBlockInfo) {
+      return Promise.all(
+        listingIds.map(async listingData => {
+          const { listingId, blockInfo } = listingData
+          return await this.getListing(listingId, blockInfo)
```

Comment:
I'm working on implementing this logic on the discovery side. Quick clarifications:
 - On line 145, why do we need to pass blockInfo when calling this.getListing ? I thought withBlockInfo options just indicates that blockInfo for the latest version of each listing should be returned along with listingId ?
 - at a higher level, what is actually the use case for the withBlockInfo options on getListings ? My understanding was that for editable listings, blockInfo always comes from data on the offer, and then we make individual calls to getListing and pass the offer blockInfo to fetch the listing at that point of time. Do we actually need the withBlockInfo option on getListings ? I could see that for optimization we could perhaps define a new function getListingsById([(listingId1, blockNum1, logIndex1), ..., (listingIdN, blockNumN, logIndexN)) to batch retrieve a list of listings as part of the getPurchases implementation. But that's separate.

Happy to jump on a call to chat if faster compared to commenting on this PR :)

Reply:
@franckc 
>On line 145, why do we need to pass blockInfo when calling this.getListing ?

The `blockInfo` object is used [here](https://github.com/OriginProtocol/origin/blob/62173514375f77b6537399f1af96ed9e5f4a0bc8/origin-js/src/contractInterface/marketplace/v00_adapter.js#L266) so to return the listing data as it was when an offer was made. This is different than the `withBlockInfo` boolean option that can be passed into `getListings()`. If you think that's too confusing or if you have an idea for how to make it better, let me know!

------------------------------------------------

>at a higher level, what is actually the use case for the withBlockInfo options on getListings ?

Yeah, you've got it right... I think this could definitely use a refactor at some point to make it less confusing, but when you call `getListings()` with `opts.purchasesFor`, it's actually fetching the `OfferCreated` events ([here](https://github.com/OriginProtocol/origin/blob/62173514375f77b6537399f1af96ed9e5f4a0bc8/origin-js/src/contractInterface/marketplace/v00_adapter.js#L385-L388)) and that's where we get the block info. 

In reality, `withBlockInfo` only works with the `PurchasesFor` option, which is even more reason for a future refactor IMO. However, this is as clean as I can figure out how to make it without really changing the scope of this PR.

Reply:
Oh dear.... Thanks for the details ! Yes, it is pretty confusing ! :)
A few ideas/comments:
 - if withBlockInfo only works in conjunction with PurchasesFor, do we need both ? could we just rely on purchasesFor ?
 - let's add comments throughout the code to help the reader follow thru this logic. otherwise I have the feeling we'll get confused every time we look at this code
 - strong unit test coverage will help us being more confident we can refactor without breaking things in the future. so the more unit test the better in my opinion :)

Reply:
> if withBlockInfo only works in conjunction with PurchasesFor, do we need both ? 

They could probably be combined, but IMO this should be left for the refactor because it changes the way the `opts` work in a fundamental way by preventing the user from getting listing ID's only (not full listing data) in certain cases, and I'd want to make sure we aren't using that and that there's not a good use case for it before we remove it... I know it's messy, but I'd rather do it as part of a refactor that cleans all of this stuff up at once and tests thoroughly that we aren't breaking anything... sound okay to you?